File: LeetCode-CPP/20.有效的括号.cpp
/*
 * @lc app=leetcode.cn id=20 lang=cpp
 *
 * [20] 有效的括号
 */

// @lc code=start
class Solution {
public:
    bool isValid(string s) {
        int n = s.size();
        unordered_map<char, char> pairs = {
            {')','('},
            {']','['},
            {'}','{'}
        };
        if(n%2 ==1){
            return false;
        }
        stack<char> stk;
        for(char ch:s){
            if(pairs.count(ch)){
                if(stk.empty() || stk.top() != pairs[ch]){
                    return false;
                }
                stk.pop();
            }
            else{
                stk.push(ch);
            }
        }
        return stk.empty();
    }
};
// @lc code=end

File: LeetCode-C++/20.有效的括号.py
#
# @lc app=leetcode.cn id=20 lang=python
#
# [20] 有效的括号
#

# @lc code=start
class Solution(object):
    def isValid(self, s):
        stack = list([])
        for ch in s:
            if ch == '(' or ch == '[' or ch == '{':
                stack.append(ch)
            else:
                if len(stack) == 0:
                    return False
                if ch == ')':
                    if stack[len(stack) - 1] != '(':
                        return False
                elif ch == ']':
                    if stack[len(stack) - 1] != '[':
                        return False
                elif ch == '}':
                    if stack[len(stack) - 1] != '{':
                        return False
                stack.pop()
        return len(stack) == 0
# @lc code=end

